File: backend/src/main/java/ch/zuehlke/hatch/sailingserver/data/MessageReceiver.java
package ch.zuehlke.hatch.sailingserver.data;

import ch.zuehlke.hatch.sailingserver.config.ApplicationConfig;
import ch.zuehlke.hatch.sailingserver.signalk.model.subscription.SignalKSubscriptionFactory;
import ch.zuehlke.hatch.sailingserver.signalk.model.subscription.SignalKSubscriptionPath;
import com.mongodb.reactivestreams.client.MongoDatabase;
import com.mongodb.reactivestreams.client.Success;
import org.bson.Document;
import org.reactivestreams.Publisher;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.socket.WebSocketMessage;
import org.springframework.web.reactive.socket.client.ReactorNettyWebSocketClient;
import org.springframework.web.reactive.socket.client.WebSocketClient;
import reactor.core.publisher.Mono;

import javax.annotation.PostConstruct;
import java.net.URI;

@Component
public class MessageReceiver {

    private final String SIGNALK_HOST = "localhost";
    private final String SIGNALK_PORT = "3000";

    private final MongoDatabase database;

    @Autowired
    public MessageReceiver(MongoDatabase database) {
        this.database = database;
    }

    @PostConstruct
    public void startReceiver() {
        WebSocketClient client = new ReactorNettyWebSocketClient();
        client.execute(
                URI.create("ws://" + SIGNALK_HOST + ":" + SIGNALK_PORT + "/signalk/v1/stream/?subscribe=none"),
                session ->
                        session
                                .send(Mono.just(session.textMessage(getInitialSubscription())))
                                .thenMany(session.receive()
                                        .map(WebSocketMessage::getPayloadAsText)
//                                        .log()
                                        .map(this::mapToDocument)
                                        .flatMap(this::store)
                                        .log()
                                )
                                .then())
                .subscribe();
    }

    private Publisher<Success> store(Document document) {
        return this.database
                .getCollection(ApplicationConfig.COLLECTION_NAME_EVENTS)
                .insertOne(document);
    }

    private Document mapToDocument(String content) {
        String sanitizedContent = content.replace("$source", "_source");

        return Document.parse(sanitizedContent);
    }

    private String getInitialSubscription() {
        return SignalKSubscriptionFactory.createDefaultSubscriptionWithSinglePath(
                "*",
                1000,
                SignalKSubscriptionPath.FORMAT_DELTA,
                SignalKSubscriptionPath.POLICY_INSTANT,
                1000).toString();
    }
}

File: backend/src/main/kotlin/ch/zuehlke/hatch/sailingserver/data/receiver/MessageReceiver.kt
package ch.zuehlke.hatch.sailingserver.data.receiver

import ch.zuehlke.hatch.sailingserver.data.eventstore.EventStore
import ch.zuehlke.hatch.sailingserver.data.eventstore.Success
import ch.zuehlke.hatch.sailingserver.signalk.model.subscription.SignalkSubscription
import ch.zuehlke.hatch.sailingserver.signalk.model.subscription.SubscriptionInfo
import com.fasterxml.jackson.databind.ObjectMapper
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import org.reactivestreams.Publisher
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.reactive.socket.client.ReactorNettyWebSocketClient
import reactor.core.publisher.Mono
import java.net.URI
import javax.annotation.PostConstruct

@Component
class MessageReceiver(private val eventStore: EventStore) {

    @Value("\${signalk.subscription.endpoint}")
    lateinit var signalkSubscriptionEndpoint: String

    private val initialSubscription = SignalkSubscription(
            "vessels.self",
            listOf(SubscriptionInfo("*", "1000", "delta", "instant", "10")))

    private val objectMapper = ObjectMapper()
    private val parser = JsonParser()

    @PostConstruct
    fun startReceiver() {
        val client = ReactorNettyWebSocketClient()
        client.execute(
                URI.create(signalkSubscriptionEndpoint)
        ) { session ->
            session
                    .send(Mono.just(session.textMessage(objectMapper.writeValueAsString(initialSubscription))))
                    .thenMany(session.receive()
                            .map { it.payloadAsText }
                            .map { this.parseJsonObject(it) }
                            .flatMap { this.store(it) }
                    )
                    .then()
        }.subscribe()
    }

    private fun store(payload: JsonObject): Publisher<Success> {
        return this. eventStore.insert(payload)
    }

    private fun parseJsonObject(payload: String): JsonObject {
        val sanitizedContent = payload.replace("\$source", "_source")

        return this.parser.parse(sanitizedContent).asJsonObject
    }
}